### PR TITLE
Fix issue 19893 - extern(C++, "ns") should count as module scope for version declarations

### PR DESCRIPTION
**Pull request by thewilsonator**
cc @TurkeyMan

### CODE REVIEW

**Comment by thewilsonator** [comment]
Ah, crap. I messed up with git.

**Comment by thewilsonator** [comment]
Hmm, doesn't work.

File: src/dmd/dversion.d [hunk 1]
```diff
@@ -199,3 +199,16 @@ extern (C++) final class VersionSymbol : Dsymbol
         v.visit(this);
     }
 }
+
+private Module moduleOf(ScopeDsymbol sds)
```
Comment (WalterBright) [inline src/dmd/dversion.d]: A function that nobody calls and has no documentation?

**Review by WalterBright — CHANGES_REQUESTED** [review]
I have no idea what this code is supposed to do.

File: src/dmd/dversion.d (hunk 1 above)
Comment (thewilsonator) [reply src/dmd/dversion.d]: The joy of buggering around with git. I'll fix this up soon.

Comment (thewilsonator) [reply src/dmd/dversion.d]: There we go.

File: src/dmd/dversion.d [hunk 2]
```diff
@@ -199,3 +199,17 @@ extern (C++) final class VersionSymbol : Dsymbol
...
```
Comment (jacob-carlborg) [inline src/dmd/dversion.d]: Would be nice with some documentation.

File: src/dmd/dversion.d [hunk 3]
```diff
@@ -199,3 +199,17 @@ extern (C++) final class VersionSymbol : Dsymbol
...
+    while (s.isNspace())
+    {
+        if (Module m = s.isModule())
+            return m;
+        s = s.parent;
+    }
+    if (Module m = s.isModule())
+        return m;
+    return null;
+}
```
Comment (jacob-carlborg) [inline src/dmd/dversion.d]: The implementation looks quite specific but the API looks quite generic.

File: src/dmd/dversion.d (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/dversion.d]: Done

File: src/dmd/dversion.d (hunk 3 above)
Comment (thewilsonator) [reply src/dmd/dversion.d]: I'm not sure what you mean by this, but I changed the function name to better reflect what it does.

**Comment by Geod24** [comment]
Please add the issue title in your commit title.

**Comment by thewilsonator** [comment]
Done.

**Comment by Geod24** [comment]
Where are those `addMembers` called from ? We have something to bypass Nspace when they are mangle only. This seems to bypass them all the time.

**Comment by thewilsonator** [comment]
> Where are those addMembers called from ?

From dsymbol semantic I presume.

> We have something to bypass Nspace when they are mangle only.

You mean https://github.com/dlang/dmd/blob/a24bed05c6f71762df462ee4ecb033c2a1ad4f49/src/dmd/dsymbol.d#L450 ?

> This seems to bypass them all the time.

This is about their location in a file not their semantic relation, e.g. one can also to 
```
module foo
extern(C++,bar):
version = baz;
```
and run into the same problem. Manu doesn't care about the non mangle only form so I presume thats why he reported the problem with the mangle only example.

**Comment by Geod24** [comment]
> You mean

Yep, although I've just checked and `mixin template` errors out.

> Manu doesn't care about the non mangle only form

No one does, but it's not a reason to wreck them. The "regular" version are supposed to introduce a scope. Maybe it would be a good time to deprecate them.

**Comment by thewilsonator** [comment]
> No one does, but it's not a reason to wreck them.

This doesn't wreck them at all, it simply allows one to declare versions in them.

> The "regular" version are supposed to introduce a scope.

Indeed they do, however you can still do `extern(C++,foo):`.

> Maybe it would be a good time to deprecate them. 

Sounds like a plan.

**Pull request closed without merging**